From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Moritz Sanft <58110325+msanft@users.noreply.github.com>
Date: Thu, 12 Feb 2026 18:52:05 +0100
Subject: [PATCH] virtcontainers: work around a kernel bug for certain TDX VM
 sizes

TDX VMs with >=64GB of memory run into a kernel panic when their memory size (in MiB) is not aligned to a multiple of 1024. To work around this, this patch rounds up the memory size to the next multiple of 1024 for the affected TDX VMs.
---
 src/runtime/virtcontainers/qemu.go | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/src/runtime/virtcontainers/qemu.go b/src/runtime/virtcontainers/qemu.go
index 3b77ff8cc2e5dcec04d172802a2a0d878b59b8b0..1b44e32f23c31b04135a758a28675afc063d030f 100644
--- a/src/runtime/virtcontainers/qemu.go
+++ b/src/runtime/virtcontainers/qemu.go
@@ -336,6 +336,19 @@ func (q *qemu) memoryTopology() (govmmQemu.Memory, error) {
 		return q.arch.memoryTopology(memMb, hostMemMb, uint8(q.config.MemSlots)), nil
 	}
 
+	// Work around the kernel bug described in [1] and [2] by aligning
+	// the memory size to a multiple of 1024 for TDX guests with >= 64
+	// Gigabyte of memory.
+	// QEMU adds a little more than 2GB to the count that's relevant
+	// for the aforementioned bug, so the mitigation is applied with a
+	// bit of a safety margin.
+	//
+	// [1]: https://lore.kernel.org/linux-efi/c2632da9-745d-46d8-901a-604008a14ac4@edgeless.systems/T/#u
+	// [2]: https://github.com/canonical/tdx/issues/421
+	if q.arch.getProtection() == tdxProtection && memMb >= 61*1024 {
+		memMb = (memMb + 1023) &^ 1023
+	}
+
 	return q.arch.memoryTopology(memMb, 0, 0), nil
 }
 
